docs/memories.md

```diff
@@ -0,0 +1,162 @@
+By defining custom memories, you can optimize your programs to target specific hardware architectures.
+
+## Overview
+
+- **Custom Memories**: Define your own memory types by subclassing the `Memory` class.
+- **Usage**: Use custom memories as annotations in your Exo code or set them during scheduling.
+
+## Defining Custom Memories
+
+To define a custom memory, you need to create a class that inherits from `Memory` and implement the required methods.
+Below is an example of defining an `AVX512` memory, which models the AVX-512 vector registers.
+
+### Example: Defining AVX512 Memory
+
+```python
+class AVX512(Memory):
+    @classmethod
+    def global_(cls):
+        return "#include <immintrin.h>"
+
+    @classmethod
+    def can_read(cls):
+        return False
+
+    @classmethod
+    def alloc(cls, new_name, prim_type, shape, srcinfo):
+        if not shape:
+            raise MemGenError(f"{srcinfo}: AVX512 vectors are not scalar values")
+        if not prim_type == "float":
+            raise MemGenError(f"{srcinfo}: AVX512 vectors must be f32 (for now)")
+        if not _is_const_size(shape[-1], 16):
+            raise MemGenError(f"{srcinfo}: AVX512 vectors must be 16-wide")
+        shape = shape[:-1]
+        if shape:
+            result = f'__m512 {new_name}[{"][".join(map(str, shape))}];'
+        else:
+            result = f"__m512 {new_name};"
+        return result
+
+    @classmethod
+    def free(cls, new_name, prim_type, shape, srcinfo):
+        return ""
+
+    @classmethod
+    def window(cls, basetyp, baseptr, indices, strides, srcinfo):
+        assert strides[-1] == "1"
+        idxs = indices[:-1] or ""
+        if idxs:
+            idxs = "[" + "][".join(idxs) + "]"
+        return f"{baseptr}{idxs}"
+```
+
+#### Explanation of Methods
+
+- **`global_(cls)`**: Returns any global code or headers needed. Here, it includes the AVX-512 intrinsic header.
+
+  ```python
+  @classmethod
+  def global_(cls):
+      return "#include <immintrin.h>"
+  ```
+
+- **`can_read(cls)`**: Controls whether the memory can be read directly. Setting it to `False` means you cannot read/write directly to this memory using standard array access.
+
+  ```python
+  @classmethod
+  def can_read(cls):
+      return False
+  ```
+
+- **`alloc(cls, new_name, prim_type, shape, srcinfo)`**: Defines how memory allocation is handled. For `AVX512`, it ensures that the allocated memory represents 16-wide vectors of `float` type.
+
+  ```python
+  @classmethod
+  def alloc(cls, new_name, prim_type, shape, srcinfo):
+      # Validation checks and allocation code
+  ```
+
+- **`free(cls, new_name, prim_type, shape, srcinfo)`**: Handles memory deallocation. For `AVX512`, no action is needed.
+
+  ```python
+  @classmethod
+  def free(cls, new_name, prim_type, shape, srcinfo):
+      return ""
+  ```
+
+- **`window(cls, basetyp, baseptr, indices, strides, srcinfo)`**: Defines how to access elements in the memory.
+
+  ```python
```

Comment:
The window member functions needs to be explained more thoroughly.

Reply:
Done!